cmd/minishift/cmd/hostfolder/add.go

```diff
@@ -20,45 +20,223 @@ import (
+
+	var name string
+	if usersShare && runtime.GOOS == "windows" {
+		// Windows-only (CIFS), all instances
+		name = "Users"
+	} else {
+		if len(args) < 1 {
+			atexit.ExitWithMessage(1, usage)
 		}
+		name = args[0]
+	}
+
+	if hostFolderManager.Exist(name) {
+		atexit.ExitWithMessage(1, fmt.Sprintf("there is already a host folder with the name '%s' defined", name))
+	}
 
-		if err != nil {
-			atexit.ExitWithMessage(1, fmt.Sprintf("Failed to mount host folder: %s", err.Error()))
+	switch shareType {
+	case hf.CIFS.String():
+		if interactive {
+			addCIFSInteractive(hostFolderManager, name)
+		} else {
+			addCIFSNonInteractive(hostFolderManager, name)
+		}
+	case hf.SSHFS.String():
+		if interactive {
+			addSSHFSInteractive(hostFolderManager, name)
+		} else {
+			addSSHFSNonInteractive(hostFolderManager, name)
 		}
-	},
+	default:
+		atexit.ExitWithMessage(1, fmt.Sprintf(unknownType, shareType))
+	}
 }
 
-func init() {
-	HostfolderCmd.AddCommand(hostfolderAddCmd)
-	hostfolderAddCmd.Flags().BoolVarP(&instanceOnly, "instance-only", "", false, "Defines the host folder only for the current OpenShift cluster.")
+func addSSHFSInteractive(manager *hf.Manager, name string) error {
+	source := util.ReadInputFromStdin("source path")
+	mountPath := readInputForMountPoint(name)
 
-	// Windows-only
-	if runtime.GOOS == "windows" {
-		hostfolderAddCmd.Flags().BoolVarP(&usersShare, "users-share", "", false, "Defines the shared Users folder as the host folder on a Windows host.")
+	config := config.HostFolderConfig{
+		Name: name,
+		Type: hf.SSHFS.String(),
+		Options: map[string]string{
+			config.Source:     source,
+			config.MountPoint: mountPath,
+		},
+	}
+	hostFolder := hf.NewSSHFSHostFolder(config, miniConfig.AllInstancesConfig)
+	manager.Add(hostFolder, !instanceOnly)
+
+	return nil
+}
+
+func addSSHFSNonInteractive(manager *hf.Manager, name string) error {
+	if source == "" {
+		atexit.ExitWithMessage(1, noSource)
+	}
+
+	if target == "" {
+		atexit.ExitWithMessage(1, noTarget)
+	}
+
+	config := config.HostFolderConfig{
+		Name: name,
+		Type: hf.SSHFS.String(),
+		Options: map[string]string{
+			config.Source:     source,
+			config.MountPoint: target,
+		},
+	}
+	hostFolder := hf.NewSSHFSHostFolder(config, miniConfig.AllInstancesConfig)
+	manager.Add(hostFolder, !instanceOnly)
+
+	return nil
+}
+
+func addCIFSInteractive(manager *hf.Manager, name string) error {
+	var uncPath string
+	if usersShare {
+		uncPath = "[determined on startup]"
+	} else {
+		uncPath = util.ReadInputFromStdin("UNC path")
+	}
+
+	if len(uncPath) == 0 {
+		return errors.New("no remote path has been specified")
+	}
+
+	mountPoint := readInputForMountPoint(name)
+
+	username := util.ReadInputFromStdin("Username")
+	if !util.IsTtySupported() {
+		return fmt.Errorf(nonSupportedTtyError)
+	}
+
+	password := util.ReadPasswordFromStdin("Password")
+	password, err := util.EncryptText(password)
+	if err != nil {
+		return err
+	}
+
+	domain := util.ReadInputFromStdin("Domain")
```

Comment:
Would also need tty support... right?

Reply:
you guys tell me. This code I jsut refaxcrtored. The original code from hostfolder.go looks like this:

```Golang
func Add(name string, allInstances bool) error {
	if isHostfolderDefinedByName(name) {
		return fmt.Errorf("Already have a host folder defined for: '%s'", name)
	}

	uncpath := util.ReadInputFromStdin("UNC path")
	if len(uncpath) == 0 {
		return fmt.Errorf("No remote path has been given")
	}
	mountpoint := readInputForMountpoint(name)
	username := util.ReadInputFromStdin("Username")
	if !util.IsTtySupported() {
		return fmt.Errorf(NonSupportedTtyError)
	}
	password := util.ReadPasswordFromStdin("Password")
	domain := util.ReadInputFromStdin("Domain")
	password, err := util.EncryptText(password)
	if err != nil {
		return err
	}

	addToConfig(newCifsHostFolder(
		name,
		uncpath,
		mountpoint,
		username, password, domain),
		allInstances)

	return nil
}
```

I find it also odd why this check is only on Username, but tbh I don't know. Happy to do as you suggest and wrap the whole interactive process into a check for TTY support. 

@praveenkumar anything on this one? I believe you might be the one who added it

Reply:
Got you

Reply:
moved the check up